### PR TITLE
[WIP][Training - Dat] - Training research advanced ruby

### PR DESCRIPTION
**Pull request by nguyen-dat-mulodo**

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -14,4 +14,5 @@
 //= require jquery_ujs
 //= require turbolinks
 //= require bootstrap-sprockets
-//= require_tree .
+//= require cable
+//= //require_tree .
```
Comment (tran-hai-mulodo) [inline app/assets/javascripts/application.js]: @nguyen-dat-mulodo 

Please remove unuse code. Thanks.

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -13,4 +23,40 @@ a:hover {
...
+    border-right: 1px solid #666666;
+    .title {
+      -webkit-transition: 0.3s;
+      transition: 0.3s;
+    }
+    .price_line {
+      text-align: right;
+      .price {
+        list-style-type: none;
+        //border: 1px solid #eee;
```
Comment (tran-hai-mulodo) [inline app/assets/stylesheets/custom.scss]: @nguyen-dat-mulodo 

Please remove unuse code. Thanks.

File: app/controllers/line_items_controller.rb [hunk 3]
```diff
@@ -66,7 +69,11 @@ def destroy
   private
     # Use callbacks to share common setup or constraints between actions.
     def set_line_item
-      @line_item = LineItem.find(params[:id])
+      # begin
```
Comment (tran-hai-mulodo) [inline app/controllers/line_items_controller.rb]: @nguyen-dat-mulodo 

Please remove unuse code. Thanks.

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -66,6 +70,15 @@ def destroy
     end
   end
 
+  def add_favorite
```
Comment (tran-hai-mulodo) [inline app/controllers/posts_controller.rb]: @nguyen-dat-mulodo 

It'll be better if you use "increment_counter". Please reference here: https://apidock.com/rails/ActiveRecord/Base/increment_counter/class

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -68,7 +70,6 @@ def set_user
 
     # Never trust parameters from the scary internet, only allow the white list through.
     def user_params
-      # params.require(:user).permit(:name, :hashed_password, :salt)
```
Comment (tran-hai-mulodo) [inline app/controllers/users_controller.rb]: @nguyen-dat-mulodo 

Pls remove unuse code. Thanks.

File: app/forms/product_form.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+class ProductForm
+  include ActiveModel::Model
+
+  attr_accessor :title, :price, :email, :description, :image_url, :category_id
+  validates :title, :description, :image_url, :presence => true
+  validates :price, :numericality => {greater_than_or_equal_to: 0.01}
+  # validates :title, :uniqueness => true
```
Comment (tran-hai-mulodo) [inline app/forms/product_form.rb]: @nguyen-dat-mulodo 

Pls remove unuse code. Thanks.

File: app/services/export_service.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    def save_file
+      @files.each do |file|
+        file.serialize "public/file_downloads/#{file.workbook.worksheets.first.name}.xlsx"
+      end
+    end
+
+    def download
+      initialize_attributes
+      return unless create_files
```
Comment (tran-hai-mulodo) [inline app/services/export_service.rb]: @nguyen-dat-mulodo 

Pls remove unuse code. Thanks.

File: app/uploaders/image_uploader.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  storage :file
+  # storage :fog
+
+  # Override the directory where uploaded files will be stored.
+  # This is a sensible default for uploaders that are meant to be mounted:
+  def store_dir
+    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
+  end
+
+  # Provide a default URL as a default if there hasn't been a file uploaded:
```
Comment (tran-hai-mulodo) [inline app/uploaders/image_uploader.rb]: @nguyen-dat-mulodo 

Pls remove unuse code. Thanks.

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (nguyen-dat-mulodo) [reply app/assets/javascripts/application.js]: fix done.
Thanks for your reviews.

File: app/controllers/posts_controller.rb (hunk 4 above)
Comment (nguyen-dat-mulodo) [reply app/controllers/posts_controller.rb]: fix done.
Thanks for your reviews.

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (nguyen-dat-mulodo) [reply app/controllers/users_controller.rb]: cái này hiện tại bản này em kiếm không thấy chỗ sửa. Anh có thể check lại nó được không ạ 
Thanks.

File: app/forms/product_form.rb (hunk 6 above)
Comment (nguyen-dat-mulodo) [reply app/forms/product_form.rb]: fix done.
Thanks for your reviews.

File: app/services/export_service.rb (hunk 7 above)
Comment (nguyen-dat-mulodo) [reply app/services/export_service.rb]: fix done.
Thanks for your reviews.

File: app/uploaders/image_uploader.rb (hunk 8 above)
Comment (nguyen-dat-mulodo) [reply app/uploaders/image_uploader.rb]: fix done.
Thanks for your reviews.

File: app/assets/stylesheets/custom.scss (hunk 2 above)
Comment (nguyen-dat-mulodo) [reply app/assets/stylesheets/custom.scss]: fix done.
Thanks for your reviews.

File: app/controllers/line_items_controller.rb (hunk 3 above)
Comment (nguyen-dat-mulodo) [reply app/controllers/line_items_controller.rb]: fix done.
Thanks for your reviews.

**Comment by tran-hai-mulodo** [comment]
@nguyen-dat-mulodo 

I checked. It's OK. So I'll merge it. Thanks.

**Pull request merged**